### PR TITLE
[SYCL][CUDA] Handle large Y/Z range dimensions.

### PR DESCRIPTION
**Pull request by mmoadeli**
The dimensions passed to sycl::range, determine the blocks per grid and threads per blocks. Currently, calculation of thread per blocks only performed for the x dimension. This means the blocks per grid for y and z dimensions passed to cuLaunchKernel, directly come from the sycl::range arguments. This can result in an error returned on calling cuLaunchKernel, when those parameters for y and z dimensions are larger than 65535.
This PR offers a simple tuning of thread per block for larger (over 65535) values of Y and Z dimensions to make the associated blocks per grid within the allowed range.

### CODE REVIEW

**Comment by bader** [comment]
Please, add `[SYCL]` title tag + 

> To a reasonable extent, title tags can be used to signify the component changed, e.g.: [PI], [CUDA], [Doc].

https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#pull-request

I also recommend linking this PR to the issue #7854 to the issue automatically when PR is merged. https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

**Comment by mmoadeli** [comment]
verify /https://github.com/intel/llvm/issues/7854

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 1]
```diff
@@ -277,18 +277,28 @@ int getAttribute(pi_device device, CUdevice_attribute attribute) {
...
+
+  threadsPerBlock[1] = (global_work_size[1] / maxGridDimY) + 1;
+  threadsPerBlock[2] = (global_work_size[2] / maxGridDimZ) + 1;
 
   PI_CHECK_ERROR(cuOccupancyMaxPotentialBlockSize(
       &minGrid, &recommendedBlockSize, kernel->get(), NULL, local_size,
       maxThreadsPerBlock[0]));
 
+  recommendedBlockSize /= (threadsPerBlock[1] * threadsPerBlock[2]);
   (void)minGrid; // Not used, avoid warnings
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:302]: Not directly related to these changes, but is there a reason to keep this around if we are not using it?

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 2]
```diff
@@ -277,18 +277,28 @@ int getAttribute(pi_device device, CUdevice_attribute attribute) {
...
-  int recommendedBlockSize, minGrid;
+  int recommendedBlockSize, minGrid, maxGridDimY, maxGridDimZ;
+
+  cuDeviceGetAttribute(&maxGridDimY, CU_DEVICE_ATTRIBUTE_MAX_GRID_DIM_Y,
+                       device->get());
+  cuDeviceGetAttribute(&maxGridDimZ, CU_DEVICE_ATTRIBUTE_MAX_GRID_DIM_Z,
+                       device->get());
+
+  threadsPerBlock[1] = (global_work_size[1] / maxGridDimY) + 1;
+  threadsPerBlock[2] = (global_work_size[2] / maxGridDimZ) + 1;
```
[the remark this answers is not in the document]
Comment (mmoadeli) [reply sycl/plugins/cuda/pi_cuda.cpp]: thanks @steffenlarsen for the comment. That's right and addressed.

**Review by steffenlarsen — APPROVED** [review]
Looks good! 🚀

**Comment by steffenlarsen** [comment]
> /verify with #7854

@mmoadeli - Was that the intended link?

**Pull request merged**